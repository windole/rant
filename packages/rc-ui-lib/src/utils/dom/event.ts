import { MouseEvent, TouchEvent } from 'react';

export const stopPropagation = (event: TouchEvent | MouseEvent | Event): void =>
  event.stopPropagation();

export function preventDefault(
  event: TouchEvent | MouseEvent | Event,
  isStopPropagation?: boolean,
): void {
  /* istanbul ignore else */
  if (typeof event.cancelable !== 'boolean' || event.cancelable) {
    event.preventDefault();
  }

  if (isStopPropagation) {
    stopPropagation(event);
  }
}
