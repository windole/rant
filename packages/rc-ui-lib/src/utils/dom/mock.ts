export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect');
  spy.mockReturnValue(rect as DOMRect);
  return () => spy.mockRestore();
}

export function mockRect(vertical?: boolean): () => void {
  return mockGetBoundingClientRect({
    width: vertical ? 0 : 100,
    height: vertical ? 100 : 0,
    top: vertical ? 0 : 100,
    right: vertical ? 0 : 100,
    left: vertical ? 100 : 0,
    bottom: vertical ? 100 : 0,
  });
}
