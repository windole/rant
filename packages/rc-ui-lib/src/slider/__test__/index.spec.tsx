import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { mockRect } from '../../utils/dom/mock';
import Slider from '..';

describe('Slider', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
    jest.restoreAllMocks();
  });

  it('shallows correctly', () => {
    wrapper = shallow(<Slider />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('min', () => {
    wrapper = shallow(<Slider min={0} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('max', () => {
    wrapper = shallow(<Slider max={100} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('step', () => {
    wrapper = shallow(<Slider step={5} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('vertical', () => {
    wrapper = shallow(<Slider vertical />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('barHeight', () => {
    wrapper = shallow(<Slider barHeight={2} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('buttonSize', () => {
    wrapper = shallow(<Slider buttonSize={20} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('activeColor', () => {
    wrapper = shallow(<Slider activeColor="#ee0a24" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('inactiveColor', () => {
    wrapper = shallow(<Slider activeColor="#e9e9e9" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('button', () => {
    wrapper = shallow(<Slider button={<div className="custom-button">button</div>} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('button is function', () => {
    wrapper = shallow(
      <Slider button={({ value }) => <div className="custom-button">{value}</div>} value={0} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render left button and right button correctly under rang mode', () => {
    wrapper = shallow(
      <Slider
        range
        leftButton={<div className="custom-button">leftButton</div>}
        rightButton={<div className="custom-button">rightButton</div>}
        value={[20, 80]}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should emit change event under rang mode', async () => {
    mockRect(false);
    const onChange = jest.fn();
    wrapper = mount(<Slider onChange={onChange} value={[20, 80]} range />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]').at(0);
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onChange).toHaveBeenCalledWith([40, 80]);
  });

  test('should emit change event after dragging button', async () => {
    mockRect(false);
    const onChange = jest.fn();
    wrapper = mount(<Slider onChange={onChange} value={0} />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onChange).toHaveBeenCalledWith(20);
  });

  test('should not emit change event when using readonly prop', async () => {
    mockRect(false);
    const onChange = jest.fn();
    wrapper = mount(<Slider onChange={onChange} value={0} readonly />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onChange).not.toHaveBeenCalled();
  });

  test('should not emit change event when using disabled prop', async () => {
    mockRect(false);
    const onChange = jest.fn();
    wrapper = mount(<Slider onChange={onChange} value={0} disabled />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onChange).not.toHaveBeenCalled();
  });

  test('should not emit change event when using reverse prop', async () => {
    mockRect(false);
    const onChange = jest.fn();
    wrapper = mount(<Slider onChange={onChange} value={0} reverse />);
    const touchStartEvent = { touches: [{ clientX: 100, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 80, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onChange).toHaveBeenCalledWith(20);
  });

  test('should emit onChangeAfter event', async () => {
    mockRect(false);
    const onChangeAfter = jest.fn();
    wrapper = mount(<Slider onChangeAfter={onChangeAfter} value={0} />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    await button.invoke('onTouchEnd')!({} as unknown as TouchEvent);
    expect(onChangeAfter).toHaveBeenCalledWith(20);
  });

  test('should emit onDragStart event', async () => {
    mockRect(false);
    const onDragStart = jest.fn();
    wrapper = mount(<Slider onDragStart={onDragStart} value={0} />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    expect(onDragStart).toHaveBeenCalledWith(touchMoveEvent, 0);
  });

  test('should emit onDragEnd event', async () => {
    mockRect(false);
    const onDragEnd = jest.fn();
    wrapper = mount(<Slider onDragEnd={onDragEnd} value={0} />);
    const touchStartEvent = { touches: [{ clientX: 0, clientY: 0 }] } as unknown as TouchEvent;
    const touchMoveEvent = {
      touches: [{ clientX: 20, clientY: 0 }],
    } as unknown as TouchEvent;
    const button = wrapper.find('[role="slider"]');
    await button.invoke('onTouchStart')!(touchStartEvent);
    await button.invoke('onTouchMove')!(touchMoveEvent);
    await button.invoke('onTouchEnd')!(touchMoveEvent);
    expect(onDragEnd).toHaveBeenCalledWith(touchMoveEvent, 20);
  });

  test('should emit click event after clicking vertical slider', async () => {
    mockRect(true);
    const onChangeAfter = jest.fn();
    wrapper = mount(<Slider onChangeAfter={onChangeAfter} value={0} vertical />);
    await wrapper.find('.rc-slider').simulate('click', { clientX: 0, clientY: 20 });
    expect(onChangeAfter).toHaveBeenCalledWith(20);
  });

  it('click when use min prop and rang mode ', async () => {
    mockRect(false);
    const onChangeAfter = jest.fn();
    wrapper = mount(<Slider range min={10} value={[20, 80]} onChangeAfter={onChangeAfter} />);
    await wrapper.find('.rc-slider').simulate('click', { clientX: 30, clientY: 0 });
    expect(onChangeAfter).toHaveBeenCalledWith([37, 80]);
  });

  it('click when using min prop and rang prop ', async () => {
    mockRect(true);
    const onChangeAfter = jest.fn();
    wrapper = mount(
      <Slider range min={10} value={[20, 80]} onChangeAfter={onChangeAfter} vertical />,
    );
    await wrapper.find('.rc-slider').simulate('click', { clientY: 30, clientX: 0 });
    expect(onChangeAfter).toHaveBeenCalledWith([37, 80]);
  });

  it('click when using min prop, rang prop and reverse prop', async () => {
    mockRect(true);
    const onChangeAfter = jest.fn();
    wrapper = mount(
      <Slider range min={10} value={[20, 80]} onChangeAfter={onChangeAfter} vertical reverse />,
    );
    await wrapper.find('.rc-slider').simulate('click', { clientY: 30, clientX: 0 });
    expect(onChangeAfter).toHaveBeenCalledWith([20, 73]);
  });

  it('click when using min prop and reverse prop', async () => {
    mockRect(false);
    const onChangeAfter = jest.fn();
    wrapper = mount(
      <Slider range min={10} value={[80, 20]} onChangeAfter={onChangeAfter} reverse />,
    );
    await wrapper.find('.rc-slider').simulate('click', { clientX: 30, clientY: 0 });
    expect(onChangeAfter).toHaveBeenCalledWith([73, 80]);
  });
});
