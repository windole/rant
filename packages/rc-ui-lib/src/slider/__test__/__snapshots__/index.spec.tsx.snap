// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider activeColor 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": "#ee0a24",
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider barHeight 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": "2px",
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider button 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="custom-button"
      >
        button
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider button is function 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "0%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="custom-button"
      >
        0
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider buttonSize 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={
          Object {
            "height": "20px",
            "width": "20px",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider inactiveColor 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": "#e9e9e9",
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider max 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider min 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider shallows correctly 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider should render left button and right button correctly under rang mode 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "20%",
        "width": "60%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--left"
      key="0"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="custom-button"
      >
        leftButton
      </div>
    </div>
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={80}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      key="1"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="custom-button"
      >
        rightButton
      </div>
    </div>
  </div>
</div>
`;

exports[`Slider step 1`] = `
<div
  className="rc-slider"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "height": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "left": "0%",
        "width": "NaN%",
      }
    }
  >
    <div
      aria-orientation="horizontal"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Slider vertical 1`] = `
<div
  className="rc-slider rc-slider--vertical"
  onClick={[Function]}
  style={
    Object {
      "background": undefined,
      "width": undefined,
    }
  }
>
  <div
    className="rc-slider__bar"
    style={
      Object {
        "background": undefined,
        "height": "NaN%",
        "top": "0%",
      }
    }
  >
    <div
      aria-orientation="vertical"
      aria-valuemax={100}
      aria-valuemin={0}
      className="rc-slider__button-wrapper rc-slider__button-wrapper--right"
      onClick={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="slider"
      tabIndex={0}
    >
      <div
        className="rc-slider__button"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;
